### ISSUE TITLE
Chapter 6: Parsing Expressions is little hard to follow

### ISSUE THREAD

**Issue by naveen17797** [munificent/craftinginterpreters#748]
I have been stuck in this chapter for a week, i didnt build the interpreter in java, i was using a different language for the task, and i couldn't understand how the left recursive parser worked in that chapter, if you could demonstrate it with a sample series of tokens which where extracted from scanner, it would be more easier to understand.

**Comment by naveen17797** [comment]
![Screenshot from 2020-06-04 08-22-01](https://user-images.githubusercontent.com/18109258/83709600-87835080-a63c-11ea-897b-9a71bc1d7978.png)

Why it has to be equality rule? cant a expression be a assignment? like

`int a = 3 * 5 / 6;`

**Comment by fengxueem** [comment]
Hi naveen17797,
When I was reading chapter, I take down every **GRAMMER RULE**, eg 
```
expression     → equality ;
```
, mentioned in the book and place them in the comment section of the source code. This helped me a lot when implementing these **RULE**s in to real pragramming code. You may try so, hope it will also be helpful for you to understand how a **Recursive descent** does its job. It's a great example of recursion algorithm and it takes time to fully master it.
For the 2nd question:

> Why it has to be equality rule? cant an expression be an assignment? 

 From my understanding, it's totally up to you, if we take a look at how **C** was designed (https://cs.wmich.edu/~gupta/teaching/cs4850/sumII06/The%20syntax%20of%20C%20in%20Backus-Naur%20form.htm), we will find an expression is, just like what's in your mind, an assignment. 
 Bob, the auther, designed **Lox** in his way. And you can turn it into your way just as simple as 
```
expression     → assignment ;
assignment    → equality ;
```
Of course, if the language is designed like so, there has to be a function to handle both assignment and equality, like
```
private Expr expression() {
    return assignment();
}

private Expr assignment() {
    Expr expr = equality();
    if (match(EQUAL)) {
        Expr value = assignment();
        .....
    }
    return expr;
}
```

**Comment by naveen17797** [comment]
Hi @fengxueem thanks for the explanation, lets say if i have a expression like this
a = b + 1;
and tokens we get from scanner would be 
['a', '=', 'b', '1']
when i call the function expression(), it eventually calls assignment().
in this token array we have first element as 'a', and second element as '=',  the match method in the book is written to match with the current token, so when you write 
`if (match(EQUAL)) `
since the array pointer is pointing to character 'a', this would fail, we wouldn't find a `EQUAL` token if we try to match against the first element, this is what confuses me, how this is actually working, explaining it with a note stating how these methods are called would help to clarify the problem, sorry i couldn't wrap my head around this problem.

**Comment by naveen17797** [comment]
Nevermind, thanks for your help, this visualization helped me to understand the function calls

https://salimsah.github.io/Automata/cfg/Bilal_RecursiveDescentParser.html#form

**Comment by daneelsan** [comment]
Hi, don't know if I should ask here but:
In challenge 3, when he says "Report that as an error, but also parse and discard a right-hand operand with the appropriate precedence.", what does it mean with the appropriate precedence? Is this suggesting that checking if the expression begins with a binary operation should be done inside each of the functions comma(), addition(), multiplication(), etc.?

**Comment by LevitatingBusinessMan** [comment]
for others, commenting down the syntax grammar before all of your parsing functions really help:
So
```RUBY
	#comparison → addition ( ( ">" | ">=" | "<" | "<=" ) addition )*
	def self.comparison
```

Also once you are writing the code and trying it in the interpeter the syntax grammar starts making sense.

@naveen17797 You could close this issue now
